cpp/ql/src/semmle/code/cpp/ir/dataflow/TaintTracking.qll

```diff
@@ -0,0 +1,189 @@
+   * instead of `DataFlow`. Use this class when taint-tracking configurations
+   * or data-flow configurations must depend on each other.
+   *
+   * See `TaintTracking::Configuration` for the full documentation.
+   */
+  abstract class Configuration2 extends DataFlow2::Configuration {
+    bindingset[this]
+    Configuration2() { any() }
+
+    /** Holds if `source` is a taint source. */
+    // overridden to provide taint-tracking specific qldoc
+    abstract override predicate isSource(DataFlow::Node source);
+
+    /** Holds if `sink` is a taint sink. */
+    // overridden to provide taint-tracking specific qldoc
+    abstract override predicate isSink(DataFlow::Node sink);
+
+    /**
+     * Holds if taint should not flow into `node`.
+     */
+    predicate isSanitizer(DataFlow::Node node) { none() }
+
+    /**
+     * Holds if the additional taint propagation step
+     * from `source` to `target` must be taken into account in the analysis.
+     * This step will only be followed if `target` is not in the `isSanitizer`
+     * predicate.
+     */
+    predicate isAdditionalTaintStep(DataFlow::Node source,
+                                    DataFlow::Node target)
+    { none() }
+
+    final override
+    predicate isBarrier(DataFlow::Node node) { isSanitizer(node) }
+
+    final override
+    predicate isAdditionalFlowStep(DataFlow::Node source, DataFlow::Node target) {
+      this.isAdditionalTaintStep(source, target)
+      or
+      localTaintStep(source, target)
+    }
+  }
+
+  /**
+   * Holds if taint propagates from `nodeFrom` to `nodeTo` in exactly one local
+   * (intra-procedural) step.
+   */
+  predicate localTaintStep(DataFlow::Node nodeFrom, DataFlow::Node nodeTo) {
+    // Taint can flow into using ordinary data flow.
+    DataFlow::localFlowStep(nodeFrom, nodeTo)
+    or
+    // Taint can flow through expressions that alter the value but preserve
+    // more than one bit of it _or_ expressions that follow data through
+    // pointer indirections.
+    not nodeTo instanceof CompareInstruction and
+    not nodeTo instanceof InvokeInstruction and
+    nodeTo.getAnOperand() = nodeFrom
+  }
+
+  /**
+   * Holds if taint may propagate from `source` to `sink` in zero or more local
+   * (intra-procedural) steps.
+   */
+  predicate localTaint(DataFlow::Node source, DataFlow::Node sink) {
+    localTaintStep*(source, sink)
+  }
+
+  /**
+   * Holds if we do not propagate taint from `fromExpr` to `toExpr`
+   * even though `toExpr` is the AST parent of `fromExpr`.
+   */
+  private predicate noParentExprFlow(Expr fromExpr, Expr toExpr) {
+    fromExpr = toExpr.(ConditionalExpr).getCondition()
+    or
+    fromExpr = toExpr.(CommaExpr).getLeftOperand()
+    or
+    fromExpr = toExpr.(AssignExpr).getLValue() // LHS of `=`
+  }
+
+  /**
+   * Holds if we do not propagate taint from a child of `e` to `e` itself.
+   */
+  private predicate noFlowFromChildExpr(Expr e) {
```

Comment:
I actually didn't realize this entire file was in this PR. It's from @rdmarsh2's initial work, and I haven't made any attempt to update it to the new API. It doesn't even compile right now. Since I wasn't planning on taint tracking support in 1.19, I'll remove this file from this PR, and we'll come back to it later.